### PR TITLE
[410140] Extract Local Variable refactoring suggests invalid names fo…

### PR DESCRIPTION
**Pull request by JanKoehnlein**
…r arrays and generic types

Change-Id: I5a03a2f914bfa628d61042ad499753e405dbe91b

### CODE REVIEW

**Comment by eclipsewebmaster** [comment]
Issue tracker reference:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=410140

**Comment by JanKoehnlein** [comment]
Jenkins xtend-ide-tests-master-latest

File: plugins/org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/refactoring/NewFeatureNameUtil.java [hunk 1]
```diff
@@ -121,10 +121,18 @@ protected String getFancyDefaultName(XExpression expression) {
...
+		return toVariableName(type);
+	}
+	
+	protected String toVariableName(LightweightTypeReference type) {
+		if(type != null) {
+			if(type.isPrimitive())
 				return type.getSimpleName().substring(0, 1);
-			else
+			else if(type.isArray()) 
+				return toVariableName(type.getComponentType()) + 's';
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/refactoring/NewFeatureNameUtil.java]: cosmetic: two dimensional arrays would get more than one 's' in the end, simple names with 's' and the end would be strange, too

Comment (szarnekow) [inline plugins/org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/refactoring/NewFeatureNameUtil.java]: Can we reuse smth from org.eclipse.xtext.common.types.xtext.ui.JdtVariableCompletions or extract common parts?

Comment (JanKoehnlein) [inline plugins/org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/refactoring/NewFeatureNameUtil.java]: IMHO, JdtVariableCompletions has a pretty cryptic API and involves StyledText. I'd rather stick with something simpler.

Comment (szarnekow) [inline plugins/org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/refactoring/NewFeatureNameUtil.java]: Yes, the API may not be optimal for this use case. On the other hand, the current Java project carries some user settings how variables are supposed to be named and we could (should) use these were possible to provide a consistent experience. Maybe you can factor the important parts into something easier to reuse?

Comment (JanKoehnlein) [inline plugins/org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/refactoring/NewFeatureNameUtil.java]: Note that the JdtVariableCompletions is about variables in Java code while NewFeatureNameUtil is about Xtend which has e.g. different keywords.

**Comment by JanKoehnlein** [comment]
New version using JdtVariableCompletions.

Advantage:
- configured prefixes/suffixes for Java variables are used

Disadvantages
-The auto-pre/suffixing takes place inside NamingConventions, as well as adding a counter, so I had to add all feature names from the current feature call scope to the excluded names (as well as all Xtend keywords). May not be the most efficient thing to do.
- Our impl now extracts 'strings' for 'List<String>', while JDT extracts 'list'. So Xtend refactoring has different proposals form JDT
- Java keywords which are not Xtend keywords will not be valid, e.g. extracting "new Goto()" will result in the default variable name 'goto1' instead of 'goto'
- JDT does not convert UPPER_CASE_UNDERSCORE to camelCase as our former impl did.

**Comment by JanKoehnlein** [comment]
Jenkins xtend-ide-tests-master-latest

**Comment by dhuebner** [comment]
Jenkins xtend-ide-tests-master-oldest

**Comment by JanKoehnlein** [comment]
Jenkins xtend-ide-tests-master-latest

**Pull request merged**